forge-ai/src/main/java/forge/ai/AiAttackController.java

```diff
@@ -511,6 +515,88 @@ public boolean apply(final Card c) {
         return notNeededAsBlockers;
     }
 
+    public void reinforceWithBanding(final Combat combat) {
+        reinforceWithBanding(combat, null);
+    }
+    public void reinforceWithBanding(final Combat combat, final Card test) {
+        CardCollection attackers = combat.getAttackers();
+        if (attackers.isEmpty()) {
+            return;
+        }
+
+        List<String> bandsWithString = Arrays.asList("Bands with Other Legendary Creatures",
+                "Bands with Other Creatures named Wolves of the Hunt",
+                "Bands with Other Dinosaurs");
+
+        List<Card> bandingCreatures = null;
+        if (test == null) {
+            bandingCreatures = CardLists.filter(myList, card -> card.hasKeyword(Keyword.BANDING) || card.hasAnyKeyword(bandsWithString));
+
+            // filter out anything that can't legally attack or is already declared as an attacker
+            bandingCreatures = CardLists.filter(bandingCreatures, card -> !combat.isAttacking(card) && CombatUtil.canAttack(card));
+
+            bandingCreatures = notNeededAsBlockers(attackers, bandingCreatures);
+        } else {
+            // Test a specific creature for Banding
+            if (test.hasKeyword(Keyword.BANDING) || test.hasAnyKeyword(bandsWithString)) {
+                bandingCreatures = new CardCollection(test);
+            }
+        }
+
+        // respect global attack constraints
+        GlobalAttackRestrictions restrict = GlobalAttackRestrictions.getGlobalRestrictions(ai, combat.getDefenders());
+        int attackMax = restrict.getMax();
+        if (attackMax >= combat.getAttackers().size()) {
+            return;
+        }
+
+        if (bandingCreatures != null) {
+            List<String> evasionKeywords = Arrays.asList("Flying", "Horsemanship", "Shadow", "Plainswalk", "Islandwalk",
+                    "Forestwalk", "Mountainwalk", "Swampwalk");
+
+            // TODO: Assign to band with the best attacker for now, but needs better logic.
+            for (Card c : bandingCreatures) {
+                Card bestBand;
+
+                if (c.getNetPower() <= 0) {
+                    // Don't band a zero power creature if there's already a banding creature in a band
```

Comment:
I think I still see a corner case here:
- if you band creature A
- next iteration you have B with power = 0
- if no other attacker you will now filter it empty
- now could crash again?

Reply:
Ah, indeed, fair enough :) I restored the NPE guard for bestAttacker